src/FixedDataTableCellGroup.js

```diff
@@ -65,74 +69,155 @@ class FixedDataTableCellGroupImpl extends React.Component {
 
   componentWillMount() {
     this._initialRender = true;
+    this._staticCellArray = [];
   }
 
   componentDidMount() {
     this._initialRender = false;
   }
 
   render() /*object*/ {
-    var props = this.props;
-    var columns = props.columns;
-    var cells = new Array(columns.length);
-    var contentWidth = sumPropWidths(columns);
+    const props = this.props;
+    const columns = props.columns;
+    const columnGroupWidth = sumPropWidths(columns);
 
-    var isColumnReordering = props.isColumnReordering && columns.reduce(function (acc, column) {
+    const isColumnReordering = props.isColumnReordering && columns.reduce(function (acc, column) {
       return acc || props.columnReorderingData.columnKey === column.props.columnKey;
     }, false);
 
-    var currentPosition = 0;
-    for (var i = 0, j = columns.length; i < j; i++) {
-      var columnProps = columns[i].props;
-      var cellTemplate = columns[i].template;
-      var recyclable = columnProps.allowCellsRecycling && !isColumnReordering;
-      if (!recyclable || (
-        currentPosition - props.left <= props.width &&
-        currentPosition - props.left + columnProps.width >= 0)) {
-        var key = columnProps.columnKey || 'cell_' + i;
-        cells[i] = this._renderCell(
-          props.rowIndex,
-          props.rowHeight,
-          columnProps,
-          cellTemplate,
-          currentPosition,
-          key,
-          contentWidth,
-          isColumnReordering
-        );
-      }
-      currentPosition += columnProps.width;
-    }
-    var style = {
+    // get list of cells
+    this._staticCellArray = this._computeVirtualizedCells({ columnGroupWidth, isColumnReordering });
+    const nonVirtualizedCells = this._computeNonVirtualizedCells({ columnGroupWidth, isColumnReordering });
+
+    const style = {
       height: props.height,
       position: 'absolute',
-      width: contentWidth,
+      width: columnGroupWidth,
       zIndex: props.zIndex,
     };
+
     FixedDataTableTranslateDOMPosition(style, -1 * DIR_SIGN * props.left, 0, this._initialRender);
 
     return (
       <div
         className={cx('fixedDataTableCellGroupLayout/cellGroup')}
-        style={style}>
-        {cells}
+        style={style}
+      >
+        {this._staticCellArray}
+        {nonVirtualizedCells}
```

Comment:
Can this ever contain both virtualized and non-virtualized cells?  Could we just have an if statement for computing these above?

Reply:
> Can this ever contain both virtualized and non-virtualized cells? Could we just have an if statement for computing these above?

Currently, no. Only one type of cells exist at a time, so `if` sounds good.

I forgot about this question: Should we support 'allowCellsRecycling' in the new API? 
(https://github.com/schrodinger/fixed-data-table-2/pull/430/files#r275406824)
Currently, every column is seen as if `allowCellsRecycling` was true when `allowColumnVirtualization` is turned on. So this means that if the user has a heavy column and it's expensive to mount/unmount it, then scrolling can keep on deleting and creating it.

Now we could get away with this by iterating and finding such kind of columns. But if the user has specified 10000 columns, then iterating through the set to find the ones with `allowCellsRecycling` turned on is expensive. I think the fix here would be to write a selector (for now it'll probably end up inside columnWidths since it already groups and iterates every column) that outputs such non-recyclable columns.

Reply:
As long as we support turning allowColumnVirtualization off, that should be fine...

Reply:
Cool, then we can brush off the selector idea until some user brings up the issue.